File: LeetCode/iter1/c++/680.valid-palindrome-ii.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	bool validPalindrome(string s) {
		bool ret = true;
		int l = 0, r = s.size() - 1;
		while (l < r) {
			if (s[l] != s[r]) {
				ret = isPalindromeHelper(s, l + 1, r) || isPalindromeHelper(s, l, r - 1);
				break;
			}

			l++;
			r--;
		}

		return ret;
	}
private:
	bool isPalindromeHelper(string s, int l, int r) {
		bool ret = true;

		while (l < r) {
			if (s[l++] != s[r--]) {
				ret = false;
				break;
			}
		}

		return ret;
	}
};

int main()
{
	Solution solution;
	cout << (solution.validPalindrome("deeee") ? "Yes\n" : "No\n");
	cout << (solution.validPalindrome("aba") ? "Yes\n" : "No\n");
	cout << (solution.validPalindrome("abca") ? "Yes\n" : "No\n");
	cout << (solution.validPalindrome("abcd") ? "Yes\n" : "No\n");

	return 0;
}

File: LeetCode/iter1/java/680.valid-palindrome-ii.java
package org.ironman.base.playground.leetcode;

public class Solution {

    public static void main(String[] args) {
        System.out.println(validPalindrome("deeee"));
        System.out.println(validPalindrome("aba"));
        System.out.println(validPalindrome("abca"));
        System.out.println(validPalindrome("abcd"));
    }

    public static boolean validPalindrome(String s) {
        boolean ret = true;
        int l = 0, r = s.length() - 1;
        while (l < r) {
            if (s.charAt(l) != s.charAt(r)) {
                ret = palindromeHelper(s, l + 1, r) || palindromeHelper(s, l, r - 1);
                break;
            }
            l++;
            r--;
        }

        return ret;
    }

    private static boolean palindromeHelper(String s, int l, int r) {
        boolean ret = true;
        while (l < r) {
            if (s.charAt(l++) != s.charAt(r--)) {
                ret = false;
                break;
            }
        }

        return ret;
    }

}